DRILL-2100: Added deleting temporary spill directories when query is …

vdiravka opened the pull request:
…finished.

jaltekruse commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java:
```diff
@@ -110,11 +111,12 @@
   private LinkedList<BatchGroup> spilledBatchGroups = Lists.newLinkedList();
   private SelectionVector4 sv4;
-  private FileSystem fs;
+  private static FileSystem fs;
```
Won't making this static and closing it cause other queries with an external sort to fail? You should be able to put this in the close() method of this class so that it only impacts the local instance. It looks like the directory created includes the operator ID, so there shouldn't be a need to keep around the spill directory until the foreman close method.

vdiravka replied:
It was my first decision to put `closeSpillFileSystem` into close(). But it in this case we can have failed query when one leaf fragment executes `close()` and delete spill folder while another leaf fragment still spooling to the disk.
That's why we need to delete the spill folder only when root fragment (foreman) is closed.
And `fs` is initialized only once that's why every instance of class uses the same FileSystem.
If I am right one drillbit can perform only one query at the same time. And another drillbit performing an external sort uses own static FileSystem instance.

jacques-n replied:
There is no reason we need to have one directory per query. Let's just have one directory per fragment and have each fragment close its own stuff out.

StevenMPhillips replied:
I agree with Jacques, except I would probably have one directory per sort operator in a fragment, rather one per fragment. This way the ExternalSort operator would be responsible for cleaning it up, and we wouldn't have to manage it at the Fragment level.

To clarify, right now the spill files for a sort go in a directory with the path:

<root spill dir>/<query_id>/<major_frag_id>/<minor_frag_id>/<operator_id>/

Instead we should make it:

<root spill dir>/<query_id>_<major_frag_id>_<minor_frag_id>_<operator_id>/

That way, the sort operator itself will be able to cleanup its own spill directory, and there won't be any empty directories left over.

jaltekruse replied:
looks like the github markdown killed part of you message, it did get through to the email they sent out, copied modified versions of the paths here:

   root spill dir / query_id / major_frag_id / minor_frag_id / operator_id/

Instead we should make it:

    root spill dir / query_id__major_frag_id__minor_frag_id__operator_id /

I don't think we need the double underscores in the path, I just did that to separate the components of the directory name.

StevenMPhillips replied:
Thanks Jason, that looks better.

jacques-n commented:
Great point on the operators. That is much cleaner than fragment level

vdiravka commented:
I made changes due to your suggestions. 
Also I deleted spaces in one level directory. Is it better?
`tmp/drill/spill/29014cb9-7eda-f4d1-f216-ce1baf83f58c_majorfragment2_minorfragment0_operator2`
Added set for deleting paths, because `drill.exec.sort.external.spill.directories` parameter can has several directories for spooling.
For the case when drillbit process is killed to delete temporary spill folders I use `fs.deleteOnExit()`.

vdiravka commented:
@jaltekruse 
Could you please review this pull request?

adeneche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java:
```diff
@@ -116,6 +119,7 @@
   private boolean first = true;
   private int targetRecordCount;
   private final String fileName;
+  private Set<Path> currSpillDirs = new TreeSet();
```
IntelliJ complains about an "unchecked assignment" here. Can you change this to `new TreeSet<>()` instead ?

vdiravka replied:
Done.

adeneche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java:
```diff
@@ -550,7 +565,15 @@ public BatchGroup mergeAndSpill(LinkedList<BatchGroup> batchGroups) throws Schem
+    try {
+        fs.deleteOnExit(currSpillPath);
+    } catch (IOException e) {
+        throw new RuntimeException(e);
```
I have some concerns about throwing an exception here:

First, does it make sense to fail the query when `deleteOnExit()` fails ? shouldn't we just log a warning. After all, if all goes well, this folder will get deleted in the close method.

Second, if this exception throws we'll be leaking memory because we'll skip clearing batchGroupList and hyperBatch

vdiravka replied:
Agree. An exception throw was replaced with:
```
// since this is meant to be used in a batches's spilling, we don't propagate the exception
final String message = "Unable to mark spill directory \"" + currSpillPath + "\" to delete on exit";
System.err.println(message + '.');
logger.warn(message, e);
```

adeneche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java:
```diff
@@ -223,7 +227,18 @@ public void close() {
+                    }
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
same concern here: does it make sense to fail a query if we fail to delete one of the spill directories ?

vdiravka replied:
Done. An exception throw was replaced with:
```
// since this is meant to be used in a batches's cleanup, we don't propagate the exception
final String message = "Unable to delete spill directory \"" + path + "\"";
System.err.println(message + '.');
logger.warn(message, e);
```

adeneche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java:
```diff
@@ -223,7 +227,21 @@ public void close() {
+            } catch (IOException e) {
+                // since this is meant to be used in a batches's cleanup, we don't propagate the exception
+                final String message = "Unable to delete spill directory \"" + path + "\"";
+                System.err.println(message + '.');
```
I don't think printing to stderr is a good idea, logging a warning should be enough

vdiravka replied:
You are right. Updated

adeneche commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java:
```diff
@@ -554,7 +572,18 @@ public BatchGroup mergeAndSpill(LinkedList<BatchGroup> batchGroups) throws Schem
+    } catch (IOException e) {
+        // since this is meant to be used in a batches's spilling, we don't propagate the exception
+        final String message = "Unable to mark spill directory \"" + currSpillPath + "\" to delete on exit";
+        System.err.println(message + '.');
```
same here

vdiravka replied:
Updated

adeneche commented:
+1

parthchandra commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/ExternalSortBatch.java:
```diff
@@ -116,6 +119,7 @@
   private boolean first = true;
   private int targetRecordCount;
   private final String fileName;
+  private Set<Path> currSpillDirs = new TreeSet<>();
```
minor - A simple list will do here.  Also consider using Guava collections.

vdiravka replied:
Agree with replacement to Guava but not sure with list. If we use list we will have the same number of paths in collection as number of temporary files.
